src/lms-ds-loader/edfi_lms_ds_loader/mssql_lms_operations.py

```diff
@@ -589,6 +633,219 @@ def soft_delete_from_production(self, table: str, source_system: str) -> None:
+            Name of the table to soft delete on, not including the `stg_`.
+        source_system: str
+            The SourceSystem currently being processed.
+        """
+
+        assert table.strip() != "", "Argument `table` cannot be whitespace"
+
+        statement = f"""
+UPDATE
+    t
+SET
+    t.DeletedAt = getdate()
+FROM
+    lms.{table} as t
+WHERE
+    t.LMSSectionIdentifier IN (
+        SELECT
+            s.LMSSectionIdentifier
+        FROM
+           lms.LMSSection as s
+        INNER JOIN
+            lms.stg_{table} as stg
+        ON
+            stg.LMSSectionSourceSystemIdentifier = s.SourceSystemIdentifier
+        AND
+            stg.SourceSystem = s.SourceSystem
+    )
+AND
+    t.LMSUserIdentifier IN (
+        SELECT
+            u.LMSUserIdentifier
+        FROM
+           lms.LMSUser as u
+        INNER JOIN
+            lms.stg_{table} as stg
+        ON
+            stg.LMSUserSourceSystemIdentifier = u.SourceSystemIdentifier
+        AND
+            stg.SourceSystem = u.SourceSystem
+    )
+AND
+    NOT EXISTS (
+        SELECT
+            1
+        FROM
+            lms.stg_{table} as stg
+        WHERE
+            t.SourceSystemIdentifier = stg.SourceSystemIdentifier
+        AND
+            t.SourceSystem = stg.SourceSystem
+    )
+AND
+    t.DeletedAt IS NULL
+AND
+    t.SourceSystem = '{source_system}'
+"""
+
+        row_count = self._exec(statement)
+        logger.debug(f"Soft-deleted {row_count} records in table `{table}`")
+
+    def soft_delete_from_production_for_assignment_and_user_relation(
+        self, table: str, source_system: str
+    ) -> None:
+        """
+        Updates production records that do not have a match in the staging table
+        by setting their `deletedat` value to the current timestamp, but
+        only for the related assignments and users in the staging table.
+
+        Parameters
+        ----------
+        table: str
+            Name of the table to soft delete on, not including the `stg_`.
+        source_system: str
+            The SourceSystem currently being processed.
+        """
+
+        assert table.strip() != "", "Argument `table` cannot be whitespace"
+
+        statement = f"""
+UPDATE
+    t
+SET
+    t.DeletedAt = getdate()
+FROM
+    lms.{table} as t
+WHERE
+    t.AssignmentIdentifier IN (
+        SELECT
+            a.AssignmentIdentifier
+        FROM
+           lms.Assignment as a
+        INNER JOIN
+            lms.stg_{table} as stg
+        ON
+            stg.AssignmentSourceSystemIdentifier = a.SourceSystemIdentifier
+        AND
+            stg.SourceSystem = a.SourceSystem
+    )
+AND
+    t.LMSUserIdentifier IN (
+        SELECT
+            u.LMSUserIdentifier
+        FROM
+           lms.LMSUser as u
+        INNER JOIN
+            lms.stg_{table} as stg
+        ON
+            stg.LMSUserSourceSystemIdentifier = u.SourceSystemIdentifier
+        AND
+            stg.SourceSystem = u.SourceSystem
+    )
```

Comment:
As with the "section-and-user" relationship, I think this check for the user identifier is unnecessary here.